src/stream/src/executor/temporal_join.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use either::Either;
+use futures::stream::{self, PollNext};
+use futures::{StreamExt, TryStreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::util::epoch::EpochPair;
+use risingwave_hummock_sdk::HummockReadEpoch;
+use risingwave_storage::table::batch_table::storage_table::StorageTable;
+use risingwave_storage::StateStore;
+
+use super::{Barrier, Executor, Message, MessageStream, StreamExecutorError, StreamExecutorResult};
+use crate::cache::ManagedLruCache;
+use crate::common::StreamChunkBuilder;
+
+pub struct TemporalJoinExecutor<S: StateStore> {
+    left: Box<dyn Executor>,
+    right: TemporalSide<S>,
+    right_stream: Box<dyn Executor>,
+    left_join_keys: Vec<usize>,
+    right_join_keys: Vec<usize>,
+    output_indices: Vec<usize>,
+    schema: Schema,
+    chunk_size: usize,
+}
+
+struct TemporalSide<S: StateStore> {
+    source: StorageTable<S>,
+    cache: ManagedLruCache<OwnedRow, Option<OwnedRow>>,
```

Comment:
To let the cache work properly, we need to call `update_epoch` for it when barriers come.

Reply:
Does `update_epoch` has any effect if there is no explicit cache eviction?
I didn't try evicting by epoch and plan to set the cache to a bounded size later on.

Reply:
IIRC, the `ManagedLruCache` can only evict by epoch.

Reply:
Don't forget to call `evict` on the cache every epoch. 🥵